NullPointerException in TableRebalancer.doRebalance() when segment is deleted during rebalance (consistent data push refreshes')

**Description:**

  When running a table rebalance on an OFFLINE table that has consistent data push and segments actively being pushed, a NullPointerException can occur in TableRebalancer.doRebalance().

  **Environment:**
  - Apache Pinot version: 1.5 (a commit from jan 14. 2026)
  - Table type: OFFLINE with consistent segment push

  **Steps to Reproduce:**
  1. Have an OFFLINE table with consistent data push 
  2. Actively push segments, start a rebalance on the table (e.g., to move it to a different server pool)
  3. While the rebalancer is waiting for EV-IS convergence, a scheduled segment refresh runs and deletes/replaces an existing segment
  4. The rebalancer detects the IdealState version change and attempts to re-check segment states
  5. NPE is thrown


  The rebalancer throws a NullPointerException and the rebalance fails. This makes it difficult to rebalance tables that have continuous ingestion without pausing the data pipeline.

  **Stack Trace:**
  java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "java.util.Map.equals(Object)" because "oldInstanceStateMap" is null
      at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
      at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
      at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1770)
      at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
      at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
      at java.base/java.lang.Thread.run(Thread.java:1583)
  Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.equals(Object)" because "oldInstanceStateMap" is null
      at org.apache.pinot.controller.helix.core.rebalance.TableRebalancer.doRebalance(TableRebalancer.java:634)
      at org.apache.pinot.controller.helix.core.rebalance.TableRebalancer.rebalance(TableRebalancer.java:197)
      at org.apache.pinot.controller.helix.core.rebalance.TableRebalanceManager.rebalanceTable(TableRebalanceManager.java:221)
      at org.apache.pinot.controller.helix.core.rebalance.TableRebalanceManager.rebalanceTable(TableRebalanceManager.java:127)
      at org.apache.pinot.controller.helix.core.rebalance.TableRebalanceManager.lambda$rebalanceTableAsync$0(TableRebalanceManager.java:158)
      at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
      ... 3 more

  **Relevant Logs:**
  INFO  Successfully updated the IdealState
  INFO  Starting EV-IS convergence check loop, 378 unique segments to monitor in current step
  INFO  ExternalView converged in 25088ms, with 0 extensions
  INFO  IdealState version changed while waiting for ExternalView to converge, re-calculating the target assignment
  INFO  Segment state changed in IdealState from: {Server_xyz_8098=ONLINE} to: null for segment: <segment_name>, re-calculating the target assignment based on the new IdealState
  ERROR Caught exception/error while rebalancing table



cc @dang-stripe

--- a/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/DefaultRebalancePreChecker.java
+++ b/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/DefaultRebalancePreChecker.java
@@ -344,7 +344,7 @@ public class DefaultRebalancePreChecker implements RebalancePreChecker {
 
   private RebalancePreCheckerResult checkRebalanceConfig(RebalanceConfig rebalanceConfig, TableConfig tableConfig,
       Map<String, Map<String, String>> currentAssignment, Map<String, Map<String, String>> targetAssignment,
-      RebalanceSummaryResult rebalanceSummaryResult) {
+      @Nullable RebalanceSummaryResult rebalanceSummaryResult) {
     List<String> warnings = new ArrayList<>();
     boolean pass = true;
     if (rebalanceConfig.isBestEfforts()) {

--- a/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/RebalancePreChecker.java
+++ b/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/RebalancePreChecker.java
@@ -43,7 +43,7 @@ public interface RebalancePreChecker {
     public PreCheckContext(String rebalanceJobId, String tableNameWithType, TableConfig tableConfig,
         Map<String, Map<String, String>> currentAssignment, Map<String, Map<String, String>> targetAssignment,
         @Nullable TableSizeReader.TableSubTypeSizeDetails tableSubTypeSizeDetails, RebalanceConfig rebalanceConfig,
-        RebalanceSummaryResult rebalanceSummaryResult) {
+        @Nullable RebalanceSummaryResult rebalanceSummaryResult) {
       _rebalanceJobId = rebalanceJobId;
       _tableNameWithType = tableNameWithType;
       _tableConfig = tableConfig;

--- a/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java
+++ b/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java
@@ -486,7 +486,10 @@ public class TableRebalancer {
     List<String> segmentsToMove = SegmentAssignmentUtils.getSegmentsToMove(currentAssignment, targetAssignment);
     Set<String> segmentsToMonitor = new HashSet<>(segmentsToMove);
 
-    long estimatedAverageSegmentSizeInBytes = summaryResult.getSegmentInfo().getEstimatedAverageSegmentSizeInBytes();
+    long estimatedAverageSegmentSizeInBytes = 0;
+    if (summaryResult != null) {
+      estimatedAverageSegmentSizeInBytes = summaryResult.getSegmentInfo().getEstimatedAverageSegmentSizeInBytes();
+    }
     Set<String> allSegmentsFromIdealState = currentAssignment.keySet();
     TableRebalanceObserver.RebalanceContext rebalanceContext = new TableRebalanceObserver.RebalanceContext(
         estimatedAverageSegmentSizeInBytes, allSegmentsFromIdealState, segmentsToMonitor);
@@ -580,7 +583,13 @@ public class TableRebalancer {
     //
     // NOTE: Monitor the segments to be moved from both the previous round and this round to ensure the moved segments
     //       in the previous round are also converged.
+    List<String> oldSegmentsToMove = segmentsToMove;
     while (true) {
+      boolean segmentsToMoveChanged = oldSegmentsToMove != segmentsToMove;
+      if (segmentsToMoveChanged) {
+        oldSegmentsToMove = segmentsToMove;
+        segmentsToMonitor.addAll(segmentsToMove);
+      }
       // Wait for ExternalView to converge before updating the next IdealState
       IdealState idealState;
       try {
@@ -600,6 +609,9 @@ public class TableRebalancer {
         return new RebalanceResult(rebalanceJobId, RebalanceResult.Status.FAILED,
             "Caught exception while waiting for ExternalView to converge: " + e, instancePartitionsMap,
             tierToInstancePartitionsMap, targetAssignment, preChecksResult, summaryResult);
+      }
+      if (segmentsToMoveChanged) {
+        segmentsToMonitor = new HashSet<>(segmentsToMove);
       }
 
       // Re-calculate the target assignment if IdealState changed while waiting for ExternalView to converge
@@ -621,7 +633,7 @@ public class TableRebalancer {
 
           // If all the segments to be moved remain unchanged (same instance state map) in the new ideal state, apply
           // the same target instance state map for these segments to the new ideal state as the target assignment
-          boolean segmentsToMoveChanged = false;
+          segmentsToMoveChanged = false;
           if (segmentAssignment instanceof BaseStrictRealtimeSegmentAssignment) {
             // For StrictRealtimeSegmentAssignment, we need to recompute the target assignment because the assignment
             // for new added segments is based on the existing assignment
@@ -652,6 +664,7 @@ public class TableRebalancer {
                       minimizeDataMovement, tableRebalanceLogger).getLeft();
               targetAssignment = segmentAssignment.rebalanceTable(currentAssignment, instancePartitionsMap, sortedTiers,
                   tierToInstancePartitionsMap, rebalanceConfig);
+              segmentsToMove = SegmentAssignmentUtils.getSegmentsToMove(currentAssignment, targetAssignment);
             } catch (Exception e) {
               onReturnFailure("Caught exception while re-calculating the target assignment, aborting the rebalance", e,
                   tableRebalanceLogger);
@@ -773,28 +786,23 @@ public class TableRebalancer {
         Preconditions.checkState(_helixDataAccessor.getBaseDataAccessor()
                 .set(idealStatePropertyKey.getPath(), idealStateRecord, expectedVersion, AccessOption.PERSISTENT),
             "Failed to update IdealState");
-        currentAssignment = nextAssignment;
         expectedVersion++;
+        currentAssignment = nextAssignment;
+        segmentsToMove = SegmentAssignmentUtils.getSegmentsToMove(currentAssignment, targetAssignment);
         // IdealState update is successful. Update the segment list as the IDEAL_STATE_CHANGE_TRIGGER should have
         // captured the newly added / deleted segments
         allSegmentsFromIdealState = currentAssignment.keySet();
         tableRebalanceLogger.info("Successfully updated the IdealState");
       } catch (ZkBadVersionException e) {
         tableRebalanceLogger.info("Version changed while updating IdealState");
-        // Since IdealState wasn't updated, rollback the stats changes made and continue. There is no need to update
-        // segmentsToMonitor either since that hasn't changed without the IdealState update
+        // Since IdealState wasn't updated, rollback the stats changes made
         _tableRebalanceObserver.onRollback();
-        continue;
       } catch (Exception e) {
         onReturnFailure("Caught exception while updating IdealState, aborting the rebalance", e, tableRebalanceLogger);
         return new RebalanceResult(rebalanceJobId, RebalanceResult.Status.FAILED,
             "Caught exception while updating IdealState: " + e, instancePartitionsMap, tierToInstancePartitionsMap,
             targetAssignment, preChecksResult, summaryResult);
       }
-
-      segmentsToMonitor = new HashSet<>(segmentsToMove);
-      segmentsToMove = SegmentAssignmentUtils.getSegmentsToMove(currentAssignment, targetAssignment);
-      segmentsToMonitor.addAll(segmentsToMove);
     }
   }
 

